src/mapping.ts

```diff
@@ -1,241 +1,403 @@
 
-  epochStakes = lastEpoch!.stakes
-  for(let i = 0; i < epochStakes.length; i++) {
-    let epochStake = EpochStake.load(epochStakes[i])
-    const epStId = crypto.keccak256(ByteArray.fromUTF8(
-      epochStake!.stakeData
-      + epochCounter!.count.toString()
-    )).toHexString()
-    epochStake!.id = epStId
+  lastEpoch!.duration = blockTimestamp.minus(lastEpoch!.startTime);
+  lastEpoch!.save();
+
+  epochStakes = lastEpoch!.stakes;
+  for (let i = 0; i < epochStakes.length; i++) {
+    let epochStake = EpochStake.load(epochStakes[i]);
+    const epStId = crypto
+      .keccak256(
+        ByteArray.fromUTF8(
+          epochStake!.stakeData + epochCounter!.count.toString()
+        )
+      )
+      .toHexString();
+    epochStake!.id = epStId;
 
     // If this stake is the one to be unstaked, decrease the amount
     if (epochStake!.stakeData == event.params.stakingProvider.toHexString()) {
-      epochStake!.amount = epochStake!.amount.minus(event.params.amount)
-      emptyStakeArrayIndex = epochStake!.amount.isZero() ? i : -1
+      epochStake!.amount = epochStake!.amount.minus(event.params.amount);
+      emptyStakeArrayIndex = epochStake!.amount.isZero() ? i : -1;
     }
 
-    epochStake!.participation = epochStake!.amount.divDecimal(totalStaked.toBigDecimal())
-    epochStake!.save()
-    epochStakes[i] = epochStake!.id
+    epochStake!.participation = epochStake!.amount.divDecimal(
+      totalStaked.toBigDecimal()
+    );
+    epochStake!.save();
+    epochStakes[i] = epochStake!.id;
   }
 
   if (emptyStakeArrayIndex >= 0) {
-    epochStakes.splice(emptyStakeArrayIndex, 1)
+    epochStakes.splice(emptyStakeArrayIndex, 1);
   }
 
-  let epoch = new Epoch(epochCounter!.count.toString())
-  epoch.startTime = blockTimestamp
-  epoch.totalStaked = totalStaked
-  epoch.stakes = epochStakes
-  epoch.save()
+  let epoch = new Epoch(epochCounter!.count.toString());
+  epoch.startTime = blockTimestamp;
+  epoch.totalStaked = totalStaked;
+  epoch.stakes = epochStakes;
+  epoch.save();
 
-  epochCounter!.count ++
-  epochCounter!.save()
+  epochCounter!.count++;
+  epochCounter!.save();
 }
 
-
 export function handleDelegateChanged(event: DelegateChanged): void {
-  let delegation = Delegation.load(event.params.delegator.toHexString())
-  if (!delegation) {
-    delegation = new Delegation(event.params.delegator.toHexString())
-  }
-  delegation.delegator = event.params.delegator
-  delegation.delegate = event.params.toDelegate
-  delegation.save()
+  const delegatee = event.params.toDelegate;
+  const stakeDelegation = getOrCreateStakeDelegation(delegatee);
+  // It's not possible to delegate voting power if a stake doesn't exist so we
+  // can force the type. The `delegator` is always a staking provider in the
+  // `TTokenStaking` contract.
+  const stake = StakeData.load(
+    event.params.delegator.toHexString()
+  ) as StakeData;
+  stake.delegatee = stakeDelegation.id;
+  stake.save();
+  stakeDelegation.save();
```

Comment:
Shouldn't `tokenholderDelegation` be updated as well? Similarly as in `handleDelegateVotesChanged`

Reply:
Good catch! Will update!

Reply:
Ahh, actually we can skip this part(in `TokenholderDelegation` and `StakeDelegation`) because the `DelegateVotesChanged` event is also emitted in the same transaction- and this event will be handled by `handleDelegateVotesChanged`. So I'm not sure which direction we should go 😆. Which option do you guys prefer and which is clearer?

Reply:
Right, that's a good point. Let's see, voting balances can change in 2 ways:
- Staked/liquid balance increases/decreases, which triggers `DelegateVotesChanged`
- Staked/liquid positions change the delegatee, which triggers `DelegateChanged` and 2 `DelegateVotesChanged` events (one due to the change on the original delegatee balance, and one due to the change on the new delegatee balance)

So, it seems that we can just use `DelegateVotesChanged` in both cases, as you suggest.

Reply:
0316b66